lib/mcp.rb

```diff
@@ -5,23 +5,30 @@
 
 require_relative "mcp/version"
 require_relative "mcp/constants"
+require_relative "mcp/autorun"
 require_relative "mcp/app"
 require_relative "mcp/server"
 require_relative "mcp/delegator"
 require_relative "mcp/client"
 
 module MCP
+  extend MCP::Autorun
+  @app_file = cleaned_caller(1).flatten.first
   class << self
     attr_reader :server
 
     def initialize_server(name:, **options)
       @server ||= Server.new(name: name, **options)
     end
+
+    def run?
+      File.expand_path($PROGRAM_NAME) == File.expand_path(@app_file) && $ERROR_INFO.nil? && $stdin.stat.readable?
+    end
   end
 
   # require 'mcp' したファイルで最後に到達したら実行されるようにするため
   # https://docs.ruby-lang.org/ja/latest/method/Kernel/m/at_exit.html
-  at_exit { server.serve(Server::StdioClientConnection.new) if $ERROR_INFO.nil? && server }
+  at_exit { server.serve(Server::StdioClientConnection.new) if run? && $ERROR_INFO.nil? && server }
```

Comment:
I think since you moved the ERROR_INFO check to run? we don't need it here anymore